frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -66,23 +67,31 @@ public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNot
   @Override
   public boolean updateUserMarkOrAddPlaceVisit(
       PlaceVisitModel placeVisit, PlaceVisitModel.UserMark newStatus) {
-    synchronized (storage) {
-      Map<String, PlaceVisitModel> placesMap = storage.get(placeVisit.tripId());
-      if (placesMap == null) {
-        Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
-        newPlaceMap.put(placeVisit.placeId(), placeVisit);
-        storage.put(placeVisit.tripId(), newPlaceMap);
-        return false;
-      }
-      if (placesMap.get(placeVisit.placeId()) != null) {
-        placesMap.put(placeVisit.placeId(),
-            placesMap.get(placeVisit.placeId()).toBuilder().setUserMark(newStatus).build());
-        return true;
-      } else {
-        placesMap.put(placeVisit.placeId(), placeVisit);
-        return false;
-      }
-    }
+
+    boolean[] alreadyInStorage = new boolean[1];
+    alreadyInStorage[0] = false;
+
+    storage.compute(placeVisit.tripId(), (tripKey, placesMap) -> {
+        if (placesMap == null) {
+          Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+          newPlaceMap.put(placeVisit.placeId(), placeVisit);
+          return newPlaceMap;
+        }
+        
+        if (placesMap.get(placeVisit.placeId()) != null) {
```

Comment:
The null-check and subsequent get-and-update checks here are not currently atomic. You'll need to perform these in a compute function, e.g.:

```java
placesMap.compute(placeVisit.placeId(), (placeId, place) -> {
  if (place != null) {
    alreadyInStorage.set(true);
    // TODO: Update existing place and return.
  } else {
    // TODO: return a new place with the correct value.
  }
  
});
```